crates/libcgroups/src/systemd/dbus_native/dbus.rs

```diff
@@ -296,20 +303,47 @@ impl DbusConnection {
             None,
         )?;
 
-        let reply = self.receive_complete_response()?;
-
-        // note that a single received response can contain multiple
-        // messages, so we must deserialize it piece by piece
         let mut ret = Vec::new();
-        let mut buf = &reply[..];
-
-        while !buf.is_empty() {
-            let mut ctr = 0;
-            let msg = Message::deserialize(&buf[ctr..], &mut ctr)?;
-            // we reset the buf, because I couldn't figure out how the adjust_counter function
-            // should should be changed to work correctly with non-zero start counter, and this solved that issue
-            buf = &buf[ctr..];
-            ret.push(msg);
+
+        // it is possible that while receiving messages, we get some extra/previous message
+        // for method calls, we need to have an error or method return type message, so
+        // we keep looping until we get either of these. see https://github.com/containers/youki/issues/2826
+        // for more detailed analysis.
+        loop {
```

Comment:
I am not sure I understand why this loop is necessary. Shouldn't `receive_complete_response` have read everything from the socket the first time?

Reply:
The way `receive_complete_response` is implemented, it loops internally to read from socket into a buffer, and if read_bytes < buffer_size, it returns the buffer. This is because dbus uses same socket for communications, and the read will block if there is no further data to read. Hence we return once the received data is less than buffer. 
Unfortunately, the dbus messages can come in multiple bursts, so we can have two messages that get delivered together, or they can have few milliseconds between them, and thus one read call might only give single message < size of buffer. We need to loop here to keep consuming messages until we find one that we are looking for, which in this case is the method_return or error type.